### PR TITLE
Testing

### LINKED ISSUE #14: Unit testing
* What is unit testing
* What are the benefits of using unit testing
* A real example of how to use unit testing

### PR DESCRIPTION
**Pull request by ebarylko**
## Summary
Added post about why I want testing in my project, and showed an example of unit, acceptance, and integration testing in my project

closes #14

### content/md/posts/2023-07-08-testing.md
* Added post about why testing would be beneficial to my project, and showed an example of each type of test (unit, integration, acceptance)

### CODE REVIEW

File: content/md/posts/2023-07-08-unit-testing.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ :toc true}
+
+
+## Overview
+
+This article covers what unit testing is, why it is useful, and an example of unit testing in a project.
+
+## What is unit testing?
+
+ Unit testing is about testing a function, using the inputs and verifying the outputs. Unlike acceptance testing which tests the result of a collection of functions interacting together, unit testing just grabs one function and sees whether it reacts correctly given certain inputs.
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: There's a space at the beginning... does markdown ignore spaces like that?
Also I think acceptance testing which `tests the application as a black box, from the user point of view`

File: content/md/posts/2023-07-08-unit-testing.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ 
+
+## Why is it useful?
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: Not sure what's the target, perhaps a better name would be `Why testing is useful?` ...

File: content/md/posts/2023-07-08-unit-testing.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ 
+One possible situation in where you could have a unit test is if you are testing the registration feature for a website. What could be tested is that given the information for a user, registering a user with that information should result in a new user appearing in the database. That user's information should match the same information provided when registering earlier.
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: Is not clear to me why a `unit test` .... the second paragraph sounds a bit more like acceptance testing ... perhaps explaining first the scenario. would be a better start ... or maybe rephrase it a bit

File: content/md/posts/2023-07-08-unit-testing.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+This test allows you to prove that user registration for the website works. Furthermore, testing individual components allows problems to be diagnosed more quickly when debugging. Knowing which components work allows the faulty components to be identified and fixed.
+
+## How do I use unit tests in tandem with accepttance testing?
+
+With acceptance testing, you test that a collection of events occur in a specific result. If you want to test every result for all event combinations, you will end up with countless tests.
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: Perhaps `all possible scenarios` is a bit more clear than `all event combinations`.

File: content/md/posts/2023-07-08-unit-testing.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+Using a combination of unit testing and acceptance testing, you can still test certain event combinations while assuring that other event combinations would not occur.
+
+For example, If I had a website where the user sees their dashboard after logging in at the home page, I could have one acceptance test check that scenario. Given a user who heads to the home page and logs in with the correct credentials, they will then see their dashboard. To be confident that users who enter incorrect credentials don't see their dashboard, I would like to have some tests that can confirm this. However, repeating the earlier acceptance test with countless variations of incorrect credentials would be a lot of effort.
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: Perhaps `countless` is a tad _strong_.

File: content/md/posts/2023-07-08-unit-testing.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+On the other hand, If I know that the user only sees their dashboard if they are authenticated, I could instead test my function which validates the user with a slew of incorrect credentials. Since viewing the dashboard depends on the result of the validation function, proving the function does not validate the user given incorrect credentials shows that an unauthenticated user cannot view their dashboard.
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: Maybe there is some kind of diagram here that could _sell_ the idea of balance ... and show which scenarios are tested with acceptance testing and which ones are tested using unit tests....

File: content/md/posts/2023-07-08-unit-testing.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+
+## A real life example
+
+I will show an example of a unit test from a document sharing site I am working on. In this test, I check whether a registered user can login with hardcoded credentials and whether their account information is updated accordingly.
+
+Here is the code for the test.
+
+```clojure
+(t/deftest registered-user-login
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: Perhaps all the re-frame code could be summarized as a function that says `setup-xxxx` or something like that, what do you think?

File: content/md/posts/2023-07-08-unit-testing.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+   (rf/dispatch [::events/login])
+   (rf-test/wait-for [::events/fetch-documents] 
+    (let [user (rf/subscribe [::subs/user])
+          name (rf/subscribe [::subs/name])]
+      (t/are [x y] (= x y)
+        :registered @user
+        "han@skywalker.com" @name)))))
+```
+
+First, I initialize my routes, setup the database, and connect to firebase. After that, I login and wait for the user's documents to be fetched. I then check whether the name of the user matches the credentials entered earlier and if the user is noted as registered.
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: Maybe a diagram here could work better... having said that... if the goal is to illustrate a unit test... maybe let's try to find a different example that doesn't include re-frame.

File: content/md/posts/2023-07-08-unit-testing.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+This test confirms that this aspect of the application works. With this I can always check to see whether logging in works. The moment it fails, I can look around and fix the issue. This is better than having this go unnoticed, for if a user can access an account with incorrect credentials, other user's information can be tampered with and taken.
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: BTW because here is more than a function, this could be considered an integration test

File: content/md/posts/2023-07-08-unit-testing.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+
+## Conclusion
+
+Unit tests serve as a way of checking the inputs and outputs of functions, allowing for their correctness to be checked. Furthermore, they can also be used with acceptance tests to show that certain scenarios will not occur if those scenarios depend on the result of a specific function.
```
Comment (amirci) [inline content/md/posts/2023-07-08-unit-testing.md]: I think we need to review a bit how are we going to illustrate why the test is important.

File: content/md/posts/2023-07-08-testing.md [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+  (:require [haggadah.subs :as subs]
+            [cljs.test :as t :include-macros true]
+            [re-frame.core :as rf]
+            [day8.re-frame.test :as rf-test]
+            [haggadah.events :as events]
+            [haggadah.core :as core]
+            [haggadah.routes :as routes]))
+```
+
+For testing that `::initialize-db` returns default-db, which is just a map of the form `{:name nil :uid nil :user :unloaded}`,
```
Comment (amirci) [inline content/md/posts/2023-07-08-testing.md:48]: NIT: Usually unit testing doesn't involve events, etc... perhaps you can extract the _init_ function and call it from the event handler, and show that function (that just returns a map) with a `deftest` and `testing`.

File: content/md/posts/2023-07-08-testing.md [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+(def admin-login-message
+  "Hello han@skywalker.com. Welcome to your dashboard. To make a new document, click the button to your right. To share and edit your existing documents, look at your documents below")
+
+(t/use-fixtures :once c/init-firebase) ;; Sets up firebase once for each test
+(t/use-fixtures :each c/with-screenshot) ;; Takes a screenshot at the end of every test
+
+(t/deftest message-test
+  (t/testing "When the admin user exists"
+    (doto driver
```
Comment (amirci) [inline content/md/posts/2023-07-08-testing.md:132]: Maybe we can replace this (just for the blog post) with a function with a nice meaningful name like `(login-and-open-dashboard-page)` or something like that...

Also perhaps remove the screenshot in the middle

File: content/md/posts/2023-07-08-testing.md [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+      (t/is (= admin-login-message actual)))))
+```
+
+In this test, I login and head to the dashboard, and then see if the welcome message has appeared. If it does not appear, an error is thrown. If it is found, I check to see if the welcome message matches the admin login message.
+
+
+
+## Future posts
+
+For future posts about testing, I will talk about new situations I had to deal with in regards to testing, or my thoughts on how tests could be made easier to understand. I will also include more posts which focus on a specific type of test, such as unit/acceptance/integration, and expand on my experience with that type of test so far.
```
Comment (amirci) [inline content/md/posts/2023-07-08-testing.md:146]: NIT: I believe a bit more succinct could be better

```
On my next posts I will describe in more detail the setup and gotchas I found with unit, integration and acceptance testing.
```

**Comment by amirci** [comment]
Please make sure you respond to comments, even if it is an answer like "done" or "fixed it" or something like that.

File: content/md/posts/2023-07-08-unit-testing.md (hunk 1 above)
Comment (ebarylko) [reply content/md/posts/2023-07-08-unit-testing.md]: I did not realize that I was leaving spaces at the beginning. I'll go and remove those.

Comment (ebarylko) [reply content/md/posts/2023-07-08-unit-testing.md]: I'll also update the description of acceptance testing so it defines it as testing the application from the user's perspective.

File: content/md/posts/2023-07-08-unit-testing.md (hunk 2 above)
Comment (ebarylko) [reply content/md/posts/2023-07-08-unit-testing.md]: I'll make the title more clear.

File: content/md/posts/2023-07-08-unit-testing.md (hunk 3 above)
Comment (ebarylko) [reply content/md/posts/2023-07-08-unit-testing.md]: I did not realize that my unit tests were actually integration tests. I'll change the scenario so it actually sounds like a unit test.

File: content/md/posts/2023-07-08-testing.md (hunk 11 above)
Comment (ebarylko) [reply content/md/posts/2023-07-08-testing.md:48]: I did not realize that. I'll change the test so it just uses functions.

File: content/md/posts/2023-07-08-testing.md (hunk 12 above)
Comment (ebarylko) [reply content/md/posts/2023-07-08-testing.md:132]: Ok. I'll take that in mind and edit all the tests so the names are more meaningful.

File: content/md/posts/2023-07-08-testing.md (hunk 13 above)
Comment (ebarylko) [reply content/md/posts/2023-07-08-testing.md:146]: Ok.
